ex1_solution/exercise_1_x_mix_drix.py

```diff
@@ -0,0 +1,123 @@
+import copy
+from enum import Enum
+from typing import List, Generator
+
+
+class PlayerToken(Enum):
+    X = 'X'
+    O = 'O'
+    EMPTY = ' '
+
+
+class XOBoard:
+    EMPTY = PlayerToken.EMPTY
+
+    def __init__(self, board: List[List[PlayerToken]]):
+        self._board = copy.deepcopy(board)
+
+    def __str__(self) -> str:
+        dict={PlayerToken.X:"X ",PlayerToken.O:"O ",PlayerToken.EMPTY:"-"}
+        text=""
+        for row in self._board:
+            for i in row:
+                text=text+dict[i]
+            text=text+"\n"
+        return text
+
+    @staticmethod
+    def empty_board():
+        return XOBoard([[XOBoard.EMPTY] * 3 for i in range(3)])
+
+    def play_move(self, row: int, column: int, player_token: PlayerToken):
+        if self._board[row][column]==PlayerToken.EMPTY and 0<row<3 and 0<column<3:
+            self._board[row][column]=player_token
+        else:
+            pass
+
+
+    def is_winning(self, player_token: PlayerToken):
+        for row in self._board:
+            if row==[player_token]*3:
```

Comment:
Avoid code duplication. Extract a variable for the expression `[player_token]*3`